### DISCUSSION TITLE
The demo seems a little sluggish

### DISCUSSION THREAD

**Discussion by lambdaq** [unpoly/unpoly#655] [General]
https://demo.unpoly.com/ 

I assume this is because the modal dialog rendering is done *after* the http calls? 

Is it possible to render layers/fragments and making http calls in parallel?

Am I wrong or how hard is it to build something like "skeleton screen" with Unpoly?

**Comment by adam12** [comment]
Feels fast to me (Canada -> Germany for traffic).

Can you record a demo of the slowness you are seeing?

**Reply by lambdaq** [reply]
@adam12  

Thanks for the attention.  I posted the video as a tweet here

https://x.com/lambdaq/status/1814205036878295524

It's not that slow, but there is a noticable delay after clicking.

I am wondering what's the recommended way of implementing "skeleton screen"? e.g. load the popup with placeholders and simultaneously load the content.

**Reply by adam12** [reply]
I'm not aware of any easy method to do the skeleton screen. Maybe Henning or someone else has seen one.

I saw one obvious delay when selecting one of the companies in the list, which had me surprised because I had presumed for something like that we would be using `[up-instant]` or `[up-preload]` to have preloaded the response. Indeed, [there is configuration](https://github.com/unpoly/unpoly-demo/blob/2e34cf2ad1271fa67cb5357f10a625eb759b5998/app/assets/javascripts/application.js#L6-L7) for those but only when `[up-follow]` is on the element.

It [seems intentional](https://github.com/unpoly/unpoly-demo/commit/a5434d5d8054de75c66d70a79cc5d62c0a19a4c1) tho, so I am not sure.

In your own application, you could try using `[up-instant]` or `[up-preload]` (or both) to reduce the perceived load time and make it feel more instantanious. Every app is different, of course.

**Reply by triskweline** [reply]
> I saw one obvious delay when selecting one of the companies in the list, which had me surprised because I had presumed for something like that we would be using [up-instant] or [up-preload] to have preloaded the response. Indeed, [there is configuration](https://github.com/unpoly/unpoly-demo/blob/2e34cf2ad1271fa67cb5357f10a625eb759b5998/app/assets/javascripts/application.js#L6-L7) for those but only when [up-follow] is on the element.
>
> It [seems intentional](https://github.com/unpoly/unpoly-demo/commit/a5434d5d8054de75c66d70a79cc5d62c0a19a4c1) tho, so I am not sure.

Oh dear, this wasn't intended. I wanted links to have explicit `[up-*]` attributes in the demo, but we actually had a lot of links that only had an `[up-layer]` attribute, but no `[up-follow]`.

The demo now uses preloading and instant links ("act on press").

**Reply by triskweline** [reply]
> I am wondering what's the recommended way of implementing "skeleton screen"? e.g. load the popup with placeholders and simultaneously load the content.

You could probably handle the [`up:link:follow`](https://unpoly.com/up:link:follow) event, immediately draw an overlay in an skeleton state, then render the actual content from the original link's URL.

I would actually be very interested in an API for optimistic rendering, but it's not super trivial once you go beyond the happy path. E.g. if a server responds with an error code, we might end up rendering a different fragment, a different layer, or keep an overlay that we would have closed with a successful response. When we render optimistically we may need to undo those changes later depending on the response.

Maybe there's a way to make these edge cases go away, or handle them in a way it's easy for the developer to revert changes. I haven't thought about this a lot.
